server/src/main/java/org/elasticsearch/gateway/TransportNodesBatchListGatewayStartedShards.java

```diff
@@ -0,0 +1,390 @@
+    TransportNodesBatchListGatewayStartedShards.NodesGatewayBatchStartedShards,
+    TransportNodesBatchListGatewayStartedShards.NodeRequest,
+    TransportNodesBatchListGatewayStartedShards.NodeGatewayBatchStartedShards> {
+
+    public static final String ACTION_NAME = "internal:gateway/local/batch_started_shards";
+    public static final ActionType<NodesGatewayBatchStartedShards> TYPE = new ActionType<>(
+        ACTION_NAME,
+        NodesGatewayBatchStartedShards::new
+    );
+
+    private final Settings settings;
+    private final NodeEnvironment nodeEnv;
+    private final IndicesService indicesService;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    @Inject
+    public TransportNodesBatchListGatewayStartedShards(
+        Settings settings,
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        NodeEnvironment env,
+        IndicesService indicesService,
+        NamedXContentRegistry namedXContentRegistry
+    ) {
+        super(
+            ACTION_NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            Request::new,
+            NodeRequest::new,
+            ThreadPool.Names.FETCH_SHARD_STARTED,
+            NodeGatewayBatchStartedShards.class
+        );
+        this.settings = settings;
+        this.nodeEnv = env;
+        this.indicesService = indicesService;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    @Override
+    protected NodeRequest newNodeRequest(Request request) {
+        return new NodeRequest(request);
+    }
+
+    @Override
+    protected NodeGatewayBatchStartedShards newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException {
+        final NodeGatewayBatchStartedShards response = new NodeGatewayBatchStartedShards(in, node);
+        assert response.getNode() == node;
+        return response;
+    }
+
+    @Override
+    protected NodesGatewayBatchStartedShards newResponse(
+        Request request,
+        List<NodeGatewayBatchStartedShards> responses,
+        List<FailedNodeException> failures
+    ) {
+        return new NodesGatewayBatchStartedShards(clusterService.getClusterName(), responses, failures);
+    }
+
+    @Override
+    protected NodeGatewayBatchStartedShards nodeOperation(NodeRequest request, Task task) {
+        NodeGatewayBatchStartedShards batchStartedShards = new NodeGatewayBatchStartedShards(clusterService.localNode());
+        try {
+            final CountDownLatch latch = new CountDownLatch(request.getShards().size());
+            for (Map.Entry<ShardId, String> entry : request.getShards().entrySet()) {
+                threadPool.executor(ThreadPool.Names.FETCH_SHARD_STARTED).execute(() -> {
+                    batchStartedShards.addStartedShard(listStartedShard(entry.getKey(), entry.getValue()));
```

Comment:
if `listStartedShard` throws a `ElasticsearchException`,  current thread `latch.countDown();` won't be executed , and the latch will never to be zero

Reply:
Good catch, during separate batch testing, I hit this issue. Now I am going to split several batchs to fetch huge number of shards. Like flush fetch request per 10k shards.